src/hotspot/share/opto/graphInvariants.cpp

```diff
@@ -0,0 +1,617 @@
+    make_helper(andd, patterns...);
+    return andd;
+  }
+
+  bool check(const Node* center, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    for (int i = 0; i < _checks.length(); ++i) {
+      if (!_checks.at(i)->check(center, steps, path, ss)) {
+        return false;
+      }
+    }
+    return true;
+  }
+  GrowableArray<Pattern*> _checks;
+};
+
+struct HasExactlyNInputs : Pattern {
+  explicit HasExactlyNInputs(uint expect_req) : _expect_req(expect_req) {}
+  bool check(const Node* center, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    if (center->req() != _expect_req) {
+      ss.print_cr("Unexpected number of input. Expected: %d. Found: %d", _expect_req, center->req());
+      for (uint i = 0; i < center->req(); ++i) {
+        Node* in = center->in(i);
+        ss.print("  %d: ", i);
+        if (in == nullptr) {
+          ss.print_cr("nullptr");
+        } else {
+          in->dump("\n", false, &ss);
+        }
+      }
+      return false;
+    }
+    return true;
+  }
+  const uint _expect_req;
+};
+
+struct HasAtLeastNInputs : Pattern {
+  explicit HasAtLeastNInputs(uint expect_req) : _expect_req(expect_req) {}
+  bool check(const Node* center, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    if (center->req() < _expect_req) {
+      ss.print_cr("Too small number of input. Expected: %d. Found: %d", _expect_req, center->req());
+      for (uint i = 0; i < center->req(); ++i) {
+        Node* in = center->in(i);
+        ss.print("  %d: ", i);
+        if (in == nullptr) {
+          ss.print_cr("nullptr");
+        } else {
+          in->dump("\n", false, &ss);
+        }
+      }
+      return false;
+    }
+    return true;
+  }
+  const uint _expect_req;
+};
+
+struct AtInput : Pattern {
+  AtInput(uint which_input, const Pattern* pattern) : _which_input(which_input), _pattern(pattern) {}
+  bool check(const Node* center, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    assert(_which_input < center->req(), "Input number is out of range");
+    if (center->in(_which_input) == nullptr) {
+      ss.print_cr("Input at index %d is nullptr.", _which_input);
+      return false;
+    }
+    bool result = _pattern->check(center->in(_which_input), steps, path, ss);
+    if (!result) {
+      steps.push(center->in(_which_input));
+      path.push(static_cast<int>(_which_input));
+    }
+    return result;
+  }
+  const uint _which_input;
+  const Pattern* const _pattern;
+};
+
+struct NodeClass : Pattern {
+  explicit NodeClass(bool (Node::*type_check)() const) : _type_check(type_check) {}
+  bool check(const Node* center, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    if (!(center->*_type_check)()) {
+      ss.print_cr("Unexpected type: %s.", center->Name());
+      return false;
+    }
+    return true;
+  }
+  bool (Node::*_type_check)() const;
+};
```

Comment:
You could probably generalize this with a callback approach. And then one concrete implentation is the one that does the type check. Just an idea.